### PR TITLE
EthernetDeviceForBridge: try all available ips

### PR DESCRIPTION
**Pull request by rhxto**
EthernetDeviceForBridge tried only the first adddress in a bridge. It now loops all the available ones and returns a device with the first valid subnet it finds.
Launchpad bug: https://bugs.launchpad.net/juju/+bug/2056365

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps
- Make a bridge with two ips: one in the juju subnets and one not. 
- Ensure the first ip is the one juju does not managed or better yet a DHCPv6 ip
- Spin up some containers on that host and ensure they all get an ip on the first subnet
## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2056365

### CODE REVIEW

**Comment by manadart** [comment]
Many thanks for your contribution. This looks like the right solution. In order to move this one forward, you'll need to sign the CLA as detailed in the instructions above.

Alternatively, if you're happy for me to cherry pick this one, I can land it for you.

In any case please mark it ready to review so we can run requisite checks.

**Comment by rhxto** [comment]
> Many thanks for your contribution. This looks like the right solution. In order to move this one forward, you'll need to sign the CLA as detailed in the instructions above.
> 
> Alternatively, if you're happy for me to cherry pick this one, I can land it for you.
> 
> In any case please mark it ready to review so we can run requisite checks.

I just didn't have this secondary address configured in Launchpad. I just added it : )
Can you retry the workflow?
Thanks

**Review by Aflynn50 — COMMENTED** [review]
The for loop needs a break when sucessful

File: state/linklayerdevices.go [hunk 1]
```diff
@@ -597,24 +597,33 @@ func (dev *LinkLayerDevice) EthernetDeviceForBridge(
...
+					// Try next address
+					continue
+				}
+				newDev.ConfigType = network.ConfigStatic
+				newDev.ProviderSubnetId = sub.ProviderId()
+				newDev.VLANTag = sub.VLANTag()
+				newDev.IsDefaultGateway = addr.IsDefaultGateway()
+				newDev.Addresses = network.ProviderAddresses{
+					network.NewMachineAddress("", network.WithCIDR(sub.CIDR())).AsProviderAddress()}
+			} else {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply state/linklayerdevices.go:623]: @manadart :arrow_double_up:

[the remark this answers is not in the document]
Comment (rhxto) [reply state/linklayerdevices.go:623]: I agree. I'll add a break statement tomorrow.

**Comment by rhxto** [comment]
Any updates on this?

**Review by nvinuesa — COMMENTED** [review]
The changes look good, but could you please add test(s) to validate the new behavior please? you can take a look at `TestEthernetDeviceForBridgeFanMTU` or `TestEthernetDeviceForBridge` in `./state/linklayerdevices_test.go` for inspiration

**Comment by SimonRichardson** [comment]
Closing for now, as there is no movement. More than welcome to reopen if you still want to work on it.

**Comment by rhxto** [comment]
Been busy with uni. I'll take a look at in in August : )

**Comment by SimonRichardson** [comment]
@rhxto Amazing, looking forward to the contribution.

**Pull request closed without merging**